course_catalog/api.py

```diff
@@ -526,3 +530,192 @@ def parse_bootcamp_json_data(bootcamp_data, force_overwrite=False):
+            parser.upload_course_image()
+        else:
+            parser.upload_all_media_to_s3(upload_master_json=True)
+
+    log.info("Digesting %s...", course_prefix)
+    digest_ocw_course(
+        parser.get_master_json(),
+        last_modified,
+        course_instance,
+        is_published,
+        course_prefix,
+    )
+    return uid, is_published
+
+
+def sync_ocw_courses(*, force_overwrite, upload_to_s3):
+    """
+    Sync OCW courses to the database
+
+    Args:
+        force_overwrite (bool): A boolean value to force the incoming course data to overwrite existing data
+        upload_to_s3 (bool): If True, upload course media to S3
+
+    Returns:
+        dict: A mapping of CourseRun.course_run_id to its is_published value
+    """
+    raw_data_bucket = boto3.resource(
+        "s3",
+        aws_access_key_id=settings.OCW_CONTENT_ACCESS_KEY,
+        aws_secret_access_key=settings.OCW_CONTENT_SECRET_ACCESS_KEY,
+    ).Bucket(name=settings.OCW_CONTENT_BUCKET_NAME)
+
+    # get all the courses prefixes we care about
+    ocw_courses = generate_course_prefix_list(raw_data_bucket)
+
+    # loop over each course
+    uids = {}
+    for course_prefix in sorted(ocw_courses):
+        try:
+            uid, is_published = sync_ocw_course(
+                course_prefix=course_prefix,
+                raw_data_bucket=raw_data_bucket,
+                force_overwrite=force_overwrite,
+                upload_to_s3=upload_to_s3,
+            )
+
+            if uid:
+                uids[uid] = is_published
+        except:  # pylint: disable=bare-except
+            log.exception("Error encountered parsing OCW json for %s", course_prefix)
+
+    return uids
+
+
+def update_course_published(runs):
+    """
+    Fix the is_published value for each course
+
+    Args:
+        runs (iterable of CourseRun): An iterable of CourseRun
+
+    Returns:
+        set[int]: A set of Course ids
+    """
+    log.info("Fixing is_published values for each course...")
+    done_courses = set()
+    for run in runs:
+        course = run.content_object
+        if course.id not in done_courses:
+            runs_published = course.course_runs.values_list("published", flat=True)
+            course.published = any(runs_published)
+            course.save()
+            done_courses.add(course.id)
+    return done_courses
+
+
+def sync_ocw_data(*, force_overwrite, upload_to_s3):
+    """
+    Sync OCW data to database
+
+    Args:
+        force_overwrite (bool): A boolean value to force the incoming course data to overwrite existing data
+        upload_to_s3 (bool): If True, upload course media to S3
+    """
+    uids = sync_ocw_courses(force_overwrite=force_overwrite, upload_to_s3=upload_to_s3)
+    course_ids = update_course_published(
+        CourseRun.objects.filter(course_run_id__in=uids.keys())
+    )
+    for course in Course.objects.filter(id__in=course_ids):
+        # Probably quicker to do this as a bulk operation
```

Comment:
I think all 3000+ OCW courses will be updated and reindexed (or deindexed) on every run (unless no uid is returned for unchanged runs - see comment above), in which case it might be worth adding bulk ES updates for both of these.

Reply:
Anecdotally it seems like the slow part is `Loading JSON` and the Elasticsearch indexing is a minor piece of it. I would like to leave it as is but I can make an issue to address later.

Reply:
👍